compiler/rustc_target/src/spec/base/cygwin.rs

```diff
@@ -0,0 +1,108 @@
+use std::borrow::Cow;
+
+use crate::spec::{
+    Cc, DebuginfoKind, LinkerFlavor, Lld, SplitDebuginfo, Target, TargetOptions, cvs,
+};
+
+#[derive(Clone, Copy)]
+pub(crate) enum CygwinVariant {
+    Cygwin,
+    MSYS2,
+}
+
+fn opts(env: CygwinVariant) -> TargetOptions {
+    let mut pre_link_args = TargetOptions::link_args(LinkerFlavor::Gnu(Cc::No, Lld::No), &[
+        // FIXME: Disable ASLR for now to fix relocation error
+        "--disable-dynamicbase",
+        "--enable-auto-image-base",
+    ]);
+    crate::spec::add_link_args(&mut pre_link_args, LinkerFlavor::Gnu(Cc::Yes, Lld::No), &[
+        // Tell GCC to avoid linker plugins, because we are not bundling
+        // them with Windows installer, and Rust does its own LTO anyways.
+        "-fno-use-linker-plugin",
+        "-Wl,--disable-dynamicbase",
+        "-Wl,--enable-auto-image-base",
+    ]);
+    // cygwin runtime lib
+    let clib = match env {
+        CygwinVariant::Cygwin => "-lcygwin",
+        CygwinVariant::MSYS2 => "-lmsys-2.0",
+    };
+    let cygwin_libs = &[clib, "-lgcc", clib, "-luser32", "-lkernel32", "-lgcc_s"];
+    let mut late_link_args =
+        TargetOptions::link_args(LinkerFlavor::Gnu(Cc::No, Lld::No), cygwin_libs);
+    crate::spec::add_link_args(
+        &mut late_link_args,
+        LinkerFlavor::Gnu(Cc::Yes, Lld::No),
+        cygwin_libs,
+    );
+    let target_env = match env {
+        CygwinVariant::Cygwin => "",
+        CygwinVariant::MSYS2 => "msys2",
+    };
+    TargetOptions {
+        os: "cygwin".into(),
+        vendor: "pc".into(),
+        env: target_env.into(),
+        // FIXME(#13846) this should be enabled for cygwin
+        function_sections: false,
+        linker: Some("gcc".into()),
+        dynamic_linking: true,
+        dll_prefix: "".into(),
+        dll_suffix: ".dll".into(),
+        exe_suffix: ".exe".into(),
+        families: cvs!["unix"],
+        is_like_windows: true,
+        allows_weak_linkage: false,
+        pre_link_args,
+        late_link_args,
+        abi_return_struct_as_int: true,
+        emit_debug_gdb_scripts: false,
+        requires_uwtable: true,
+        eh_frame_header: false,
+        // FIXME(davidtwco): Support Split DWARF on Cygwin - may require LLVM changes to
+        // output DWO, despite using DWARF, doesn't use ELF..
+        debuginfo_kind: DebuginfoKind::Pdb,
```

Comment:
Hrm. This looks like it's using PDB, not Dwarf? I don't understand this comment. I know it's a holdover from the other Windows GNU targets, but...

Reply:
The original PR: https://github.com/rust-lang/rust/pull/98051

Actually the debug info of MinGW and Cygwin is Dwarf, but it's always merged into the executable and could not be splitted. I'm not sure that the msvc debugger could recognize dwarf, but lldb can handle both dwarf and pdb. I guess that here it just tries to keep the original behavior on Windows, and let llvm to choose the proper debug info kind. In my test, the mingw and cygwin targets still use dwarf even they're set to pdb.

Reply:
Modern Windbg can handle DWARF.

This line is the same for `*-windows-gnu*` targets.  I never understood it, but hey, as long as it works…